lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -242,54 +195,111 @@ public void OnNext(IActiveContext activeContext)
         /// <param name="completedTask">The link to the completed task</param>
         public void OnNext(ICompletedTask completedTask)
         {
-            _completedTasks.Add(completedTask);
-
-            if (_completedTasks.Count != _dataSet.Count + 1)
-            {
-                return;
-            }
-            
-            foreach (var task in _completedTasks)
+            lock (_lock)
             {
-                Logger.Log(Level.Verbose, string.Format("Disposing task: {0}", task.Id));
-                task.ActiveContext.Dispose();
+                Logger.Log(Level.Info,
+                    string.Format("Received completed task message from task Id: {0}", completedTask.Id));
+                _completedTasks.Add(completedTask);
```

Comment:
No. You are probably wondering why am I collecting them right? So at the end of IMRU master sends done message to slaves/mappers ad then exits the task. I do not want the streams to get closed unless slaves actually receive the message and also reach completed task stage.